File: src/content/cre/guides/workflow/using-triggers/http-trigger/code-snippets/simulation-example.go
//go:build wasip1

package main

import (
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/smartcontractkit/cre-sdk-go/capabilities/networking/http"
	"github.com/smartcontractkit/cre-sdk-go/cre"
	"github.com/smartcontractkit/cre-sdk-go/cre/wasm"
)

type Config struct {
	MinimumAmount int `json:"minimumAmount"`
}

type RequestData struct {
	UserID string `json:"userId"`
	Action string `json:"action"`
	Amount *int   `json:"amount"` // Pointer to detect missing field
}

func onHttpTrigger(config *Config, runtime cre.Runtime, payload *http.Payload) (string, error) {
	var data RequestData
	if err := json.Unmarshal(payload.Input, &data); err != nil {
		return "", fmt.Errorf("failed to parse input: %w", err)
	}

	// Validate required fields
	if data.UserID == "" || data.Action == "" || data.Amount == nil {
		runtime.Logger().Info("Missing required fields")
		return "Error: Missing required fields (userId, action, amount)", nil
	}

	runtime.Logger().Info("Processing request", "action", data.Action, "userId", data.UserID)

	if *data.Amount < config.MinimumAmount {
		runtime.Logger().Info("Amount below minimum", "amount", *data.Amount, "minimum", config.MinimumAmount)
		return "Amount too low", nil
	}

	runtime.Logger().Info("Processing amount", "amount", *data.Amount)
	return fmt.Sprintf("Successfully processed %s", data.Action), nil
}

func InitWorkflow(config *Config, logger *slog.Logger, secretsProvider cre.SecretsProvider) (cre.Workflow[*Config], error) {
	httpTrigger := http.Trigger(&http.Config{}) // Empty config OK for simulation

	return cre.Workflow[*Config]{
		cre.Handler(httpTrigger, onHttpTrigger),
	}, nil
}

func main() {
	wasm.NewRunner(cre.ParseJSON[Config]).Run(InitWorkflow)
}

File: src/content/cre/guides/workflow/using-triggers/http-trigger/code-snippets/simulation-example.ts
import { HTTPCapability, handler, type Runtime, type HTTPPayload, Runner, decodeJson } from "@chainlink/cre-sdk"

type Config = {
  minimumAmount: number
}

type RequestData = {
  userId: string
  action: string
  amount: number
}

const onHttpTrigger = (runtime: Runtime<Config>, payload: HTTPPayload): string => {
  const requestData = decodeJson(payload.input) as RequestData

  // Validate required fields
  if (!requestData.userId || !requestData.action || requestData.amount === undefined) {
    runtime.log("Missing required fields")
    return "Error: Missing required fields (userId, action, amount)"
  }

  runtime.log(`Processing ${requestData.action} for user ${requestData.userId}`)

  if (requestData.amount < runtime.config.minimumAmount) {
    runtime.log(`Amount ${requestData.amount} below minimum ${runtime.config.minimumAmount}`)
    return "Amount too low"
  }

  runtime.log(`Processing amount: ${requestData.amount}`)
  return `Successfully processed ${requestData.action}`
}

const initWorkflow = (config: Config) => {
  const http = new HTTPCapability()

  return [
    handler(http.trigger({}), onHttpTrigger), // Empty config OK for simulation
  ]
}

export async function main() {
  const runner = await Runner.newRunner<Config>()
  await runner.run(initWorkflow)
}
